### PR TITLE
Implement user signup

### LINKED ISSUE #465: Enable user signup
Because we're doing passwordless login, signing up will look pretty similar to signing in from a user's perspective, i.e. users will enter their email address and then receive an email with a signed link that will allow them to sign in to the app. However, handling will need to be different on the backend when the user doesn't already exist in the database. Tasks for this card:
- Create the necessary database objects when a user attempts to sign in who doesn't already exist in the system and send them an email
- Make adjustments to the frontend and/or email text to clarify to users that they are going through a signup process.

## Notes
- It will probably be helpful to be able to differentiate between users who have attempted to sign up but haven't actually clicked the email link, and users who have made it all the way to logging in to the app. This could be accomplished by not creating a `UserProfile` object until the user successfully logs in, or by storing one or more fields on the `UserProfile` (e.g. `is_verified` or `last_login_date`).

### PR DESCRIPTION
**Pull request by rachelekm**
## Overview

This PR allows for users to sign up for an account with their email address and, on form submit, receive an emailed link to create a profile and sign in. Below the surface, when a user clicks the “create account” button: their email is validated, a new `User` is created, the new `User` is added to the `HouseSeeker` group, and an empty related `UserProfile` is created before sending the email. Appropriate frontend changes were made, including:
- Adding a clickable prompt on the “sign in” page to create an account (and from the “sign up” page a prompt to toggle back to “sign in”). It should be noted that these are not two separate pages/components, but just toggles between appropriate text/actions.
- Updated `loginMessage` payloads to walk through the signup process or give context to any signup errors.

### Notes

- One tricky part of this PR was figuring out how to let a new `User` create their initial `UserProfile`, even though the authorization to reach the `EditProfile` component relies on the presence of a `UserProfile`. The way I chose to get around this, while also differentiating `Users` who have attempted to sign up but haven't actually clicked the email link or who haven't create a `UserProfile`, was to ~~create an intermediary `AccountProfile` type: `{unverifiedUserProfile: boolean}`. Responding with this object from the backend and setting it as the profile seemed simpler than creating a field on the `UserProfile` model and allows `profile !== undefined` while also differentiating between a complete `UserProfile` and `isAnonymous`~~ create an empty `UserProfile` on signup. This allows `profile !== undefined` to pass authentication and will populate/view a default `EditProfile` component as if `isAnonymous` without actually being `isAnonymous`.
- I made a couple adjustments to the `AccountProfile` type as well as some of the conditionals in `EditProfile` to better fit when a new `User` _has_ a profile, but doesn't actually have anything in it to validate. This is namely getting rid of the `key` property since this is a leftover of our Amplify/S3 profile process (I couldn't find any active purpose to it) as well as making `voucherNumber` optional. These changes were also made as part of the open PR #497 so I tried to stay as close as possible to those changes to avoid any conflicts.
- Full account creation depends on the ability to edit a `UserProfile`. However, since there’s an open PR #497 for editing `UserProfile` from the frontend I decided not to touch any of the functionality. This PR simply creates an empty `UserProfile` for new `Users` and, once we have the ability to save edits to `UserProfile`, the absence of the `unverifiedUserProfile` property will signify account creation is "complete".
- I left the email body copy the same as login since it seemed broad enough to cover signup as well. Any extra signup context is included as part of the `loginMessage` after submitting email.

## Testing Instructions

 * `scripts/update` 
 * `scripts/manage test users.tests` to confirm tests function
 * `scripts/server`
 * Navigate to `localhost:9966/`
 * Click the "create an account" text to toggle to the signup form
 * Try and submit an invalid email and confirm account creation fails and error response message appears
 * Try and submit a valid email and confirm email link prints in the console
 * Try and submit the same email again and confirm duplicate account creation fails
 * Navigate to the email link and confirm it loads `/profile` page, and all fields are present but empty
 * Confirm you can't navigate beyond `/profile` without setting required fields
 * Create a new `User` through the admin, and don't create a related `UserProfile`. Login with the new `User` email and confirm still authorized to navigate to the `/profile` page.

Resolves #465

### CODE REVIEW

File: src/backend/users/views.py
Comment (rachelekm) [inline src/backend/users/views.py:207]:
```suggestion
        try:
            with transaction.atomic():
                user_serializer = HouseSeekerSignUpSerializer(data=request.data)
```

I'm realizing after the fact I don't have a good way to rollback any database transactions in the event something happens between User creation --> add to HouseSeekers --> UserProfile creation within the serializer. Does adding something like the above make sense here?

Comment (rachelekm) [inline src/backend/users/views.py:133]:
```suggestion
        try:
            with transaction.atomic():
                user_serializer = HouseSeekerSignUpSerializer(data=request.data)
```

I'm realizing after the fact I don't have a good way to rollback any database transactions in the event something happens between User creation --> add to HouseSeekers --> UserProfile creation within the serializer. Does adding something like the above make sense here?

File: src/backend/users/serializers.py
Comment (ddohler) [inline src/backend/users/serializers.py:52]: For operations that affect multiple objects at once, it's good practice to use [`@transaction.atomic`](https://docs.djangoproject.com/en/4.0/topics/db/transactions/#django.db.transaction.atomic) in order to ensure that things don't get left in a half-completed state if something goes wrong partway through.

File: src/backend/users/tests.py
Comment (ddohler) [inline src/backend/users/tests.py]: It took me a bit to figure out what was going on here, I think because the `doesntexist` is kind of hidden in the middle of this email, maybe these could get changed to `exists@azavea.com` and `doesntexist@azavea.com`?

File: src/backend/users/tests.py
Comment (ddohler) [inline src/backend/users/tests.py]: `len(mail.outbox)` is short enough and descriptive enough on its own that I think you could probably just inline it in the next line and avoid creating a separate variable (here and in other places).

File: src/backend/users/views.py
Comment (ddohler) [reply src/backend/users/views.py:207]: Hah, yes! I made a comment about this in another spot before seeing this.

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]: TODO: What's up with this?

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]: This will have the effect of hiding some information because it'll transform the `User.DoesNotExist` exception (which is pretty descriptive) into a generic `Exception` (albeit one with a descriptive message). I think it'd be preferable to allow the User.DoesNotExist to bubble up so that it can be handled appropriately by the calling function. So I'd remove these last two lines and add a docstring noting that this function will raise an exception if the user doesn't exist, and then alter the login and signup views appropriately.

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]:
```python
except Exception:
    pass
```
Can be kind of a dangerous thing to have in a file because it'll catch and attempt to ignore pretty much any conceivable problem that could occur in the `try` block. This can make it harder to learn about and diagnose issues because there won't be any logging information available. For this reason I'd revert this to `except User.DoesNotExist` so that only that error is ignored and allow any other exceptions to bubble up so that it yells at us if there are other problems beyond what we specifically know we want to ignore.

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]: It's possible to have multiple `except` clauses, so this can be restructured as
```python
except IntegrityError:
    ...
except ValidationError:
    ...
except Exception:
    ...
```

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]: I would leave this catch-all error handling to the frontend; this catch-all will have the effect of converting all errors into 400 errors, which could be misleading to both users and ourselves. It'll also prevent stack traces from getting printed to logs which will make it harder for us to diagnose an error in production.
```suggestion

```

This means we'll need to attempt to parse the error message on the frontend, but if it doesn't exist we'll have to fall back to printing a generic error there.

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]: I suspect you may have inadvertently solved this problem twice, because it doesn't look to me like this can possibly happen. For users who come in through the ordinary flow, `HouseSeekerSignUpSerializer` will ensure that they get a `UserProfile`, so they shouldn't ever encounter this case. When I sign up with a new email address, on logging in for the first time I get a full-fledged (albeit empty) Profile object, rather than this response. Please let me know if I'm missing something here!

File: taui/src/components/edit-profile.js
Comment (ddohler) [inline taui/src/components/edit-profile.js]: As noted above, I suspect this may no longer be necessary.

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py]: I just had a thought: could we check for a user's profile and give them a blank one if it doesn't exist at this point? You're already doing that in the signup serializer, and I think it's a good idea to continue doing that, but doing it here would address the issue where someone who has been signed up manually using the admin or CLI may not have a profile set. There's no reason that we wouldn't want those people to be able to log in normally, so we might as well just give them a blank profile if we detect they're missing one.

**Review by ddohler — CHANGES_REQUESTED** [review]
There are some code style things that I think should be addressed before merging, but the overall structure of this is looking good! Everything worked as expected in my testing.

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: This is a great point! Actually, what do you think about creating a custom `User` manager class that creates an empty `UserProfile` instance with the `User` instance so everything's good to go on `create`? It might mean we have a build up of empty `UserProfile` instances, but like you said it would allow every `User` to be authorized to view the site by default once they have the link. I _think_ `Counselors` can only view/edit a `HouseSeeker` `UserProfile` in admin, but if not I can open an issue so the empty `UserProfiles` don't show up for them. Either way, I think a check for profile before sending the link is good here and will pop that in!

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py:207]: Awesome! I see your comment below and will edit this to use the @transaction.atomic decorator

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: Is this in reference to the TODO below or the comment on line 21? I added the comment to add some clarification behind how we get `email` from the `request.data`, but I'm realizing that may be more confusing/unnecessary than helpful! I'll take it out, if so.

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: Oh jeez, ok will definitely revert this back.

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: Ah, ok got it will do!

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: Another "oh jeez", ok I'll get rid of this catch-all and make sure I update the frontend appropriately.

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: Oh, huh! Nope you're right, I thought I was covering my bases but this isn't needed! I'll take this out along with the rest of the references to `unverifiedUserProfile`.

File: src/backend/users/views.py
Comment (ddohler) [reply src/backend/users/views.py]: Ugh, this was a TODO for myself to look into this closer, which I lost track of before submitting the review. Sorry! Anyway, this piqued my interest because it made me wonder whether the request payload could be different depending on whether this was a signup or a signin situation, and whether maybe we should find a way to unify them to make sure it's always consistent. But I didn't actually check into any of those things, so I don't know whether any of those concerns are valid.

File: src/backend/users/views.py
Comment (ddohler) [reply src/backend/users/views.py]: That's a good thought! If doing that, we'd need to keep in mind that there are several different ways to create objects in Django and some of them [bypass the usual methods](https://docs.djangoproject.com/en/4.0/ref/models/querysets/#bulk-create). We've put a fair bit of effort on past projects into ensuring that related objects always, _always_ get created at the same time, and it can be a fair bit of work to cover all the bases. We don't have much more work to do on this phase of the project, so I'm thinking that, for now, it's probably safe to do this in the few places where we know we'll need it, but if we add more User-related features later or find that we're encountering bugs related to profiles being missing then we should revisit and consider doing something like a custom User manager.

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: I took another look back and I think unifying the payloads is doable! When creating a new `User` it expects to find the `username` field, but I don't think there's any reason not to update the payload key from `email` to `username` when logging in. I gave it a try and logging in and signing up both are functioning as expected so far.

File: src/backend/users/views.py
Comment (rachelekm) [reply src/backend/users/views.py]: Ok, got it! For now I'll implement the check/creating an empty `UserProfile` here and will keep an eye out for any funny business testing with admin/CLI.

**Comment by rachelekm** [comment]
Thanks @ddohler! I implemented the changes from our discussion above and updated the testing instructions so this should be good to go for a second review.

File: src/backend/users/serializers.py
Comment (ddohler) [reply src/backend/users/serializers.py:52]: Looks like this still needs the `@transaction.atomic` decorator.

File: src/backend/users/views.py
Comment (ddohler) [inline src/backend/users/views.py:205]: This can create challenges for translation because these are user-visible strings that don't pass through the frontend translation framework. I wouldn't change anything now in the interest of time, but for the future, if there's translation happening client-side, I'd have the backend send back message keys, maybe something like `'SignIn.AccountExists'` or `'SignIn.InvalidEmail'`, and then use those to determine what to display to the user client-side.

**Review by ddohler — APPROVED** [review]
Works great! I think one of the comments from the previous review round slipped through without getting addressed, but aside from that this is good to go!

I did notice one problem with logging out that may have been obscured by the earlier problems with loading routing networks, or it may have been something introduced by a merge conflict. If I try to log out from `/map` I don't make it back to the login screen, and instead there's a console error saying `Uncaught (in promise) TypeError: profile is null`. If you're able to replicate this can you open an issue for it (I have been having some troubles with cached assets acting strangely, I think. I'm fairly certain this behavior is real but want someone else to double-check)?

**Comment by rachelekm** [comment]
Thanks @ddohler! I tested it locally and noticed the same problem logging out, I just opened an issue for it #511.

> one of the comments from the previous review round slipped through without getting addressed, but aside from that this is good to go!
>
^ For the `@transaction.atomic` decorator I added `with transaction.atomic():` to the `try` block in the `SignUpPage` view to allow us to customize the signup message sent back in the response if there is an `IntegrityError`. Should this instead be a decorator on the serializer's `create`?

**Comment by ddohler** [comment]
> For the @transaction.atomic decorator I added `with transaction.atomic()`: to the try block in the SignUpPage view to allow us to customize the signup message sent back in the response if there is an IntegrityError. Should this instead be a decorator on the serializer's create?

Oh, sorry, I missed that! (Though I thought I searched for it, but anyway 🤷 ) But yes, it should be moved to the serializer's `create()` method -- you want the `transaction.atomic` to be as close to any database-altering calls (`.save()`, `Model.objects.create()`, etc.) as possible so that it's guaranteed to wrap them no matter what happens outside the function. Similarly, because you [shouldn't catch exceptions inside a `transaction.atomic()`](https://docs.djangoproject.com/en/4.0/topics/db/transactions/#django.db.transaction.atomic), it's also a good idea to keep the `transaction.atomic()` as tight around the affected code as possible, which may mean using a `with transaction.atomic():` instead of the decorator if the function is larger. But in this case, you need to cover pretty much the whole `create()` function (except the `return`), so the decorator is appropriate.

**Comment by rachelekm** [comment]
Ah, ok thank you! I understand why the decorator on the serializer works better here and that change has been addressed.

**Pull request merged**
